common-lib/src/main/scala/com/gu/workflow/api/StubAPI.scala

```diff
@@ -0,0 +1,283 @@
+package com.gu.workflow.api
+
+import com.gu.workflow.api.ApiUtils._
+import com.gu.workflow.lib.{Caching, ContentAPI, QueryString}
+import io.circe.Json
+import io.circe.syntax._
+import models.DateFormat._
+import models.{ContentItemIds, ExternalData, Flag, Stub}
+import models.api._
+import org.joda.time.DateTime
+import scalacache.memoization._
+import concurrent.duration._
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+object StubAPI extends Caching {
+
+  def createStub(body: Json): ApiResponseFt[ContentUpdate] =
+    for {
+      convertedJson <- flatStubJsonToStubJson(body)
+      res <- ApiResponseFt.Async.Right(postRequest("stubs", convertedJson))
+      json <- parseBody(res.body)
+      createRes <- extractDataResponse[ContentUpdate](json)
+    } yield createRes
+
+  def putStub(stubId: Long, body: Json): ApiResponseFt[ContentUpdate] =
+    for {
+      convertedJson <- flatStubJsonToStubWithCollaboratorsJson(body)
+      res <- ApiResponseFt.Async.Right(putRequest(s"stubs/$stubId", convertedJson))
+      json <- parseBody(res.body)
+      putRes <- extractDataResponse[ContentUpdate](json)
+    } yield putRes
+
+  def getStub(id: Long): ApiResponseFt[Option[Stub]] =
+    for {
+      res <- ApiResponseFt.Async.Right(getRequest(s"stubs/$id"))
+      json <- parseBody(res.body)
+      item <- extractDataResponseOpt[Stub](json)
+    } yield item
+
+  private def getPlannedPrintLocationDescriptionFromStub(
+                                           contentAPI: ContentAPI,
+                                           stub: Option[Stub]
+                                         ): Future[Option[String]] = {
+    stub
+      .flatMap(s => Some(s.plannedPublicationId, s.plannedBookSectionId, s.plannedBookSectionId)) match {
+      case Some((Some(pId), Some(bId), Some(bsId))) => getPrintLocationDescriptionFromIds(contentAPI, pId, bId, bsId)
+      case _ => Future.successful(None)
+    }}
+
+  private def getActualPrintLocationDescriptionFromExternalData(
+                                           contentAPI: ContentAPI,
+                                           externalData: Option[ExternalData]
+                                         ): Future[Option[String]] = {
+    externalData
+      .flatMap(e => Some(e.actualPublicationId, e.actualBookId, e.actualBookSectionId)) match {
+      case Some((Some(pId), Some(bId), Some(bsId))) => getPrintLocationDescriptionFromIds(contentAPI, pId, bId, bsId)
+      case _ => Future.successful(None)
+    }}
+
+  private def getPrintLocationDescriptionFromIds(contentAPI: ContentAPI, pId: Long, bId: Long, bsId: Long): Future[Option[String]] =
+    memoize(60.minutes) {
```

Comment:
Discussed over chat but for posterity – memoizing at this level will produce a new cache key for every new set of arguments, even if some of those arguments have been seen within the cache window. For example, calls to the parent with pId, bId and bsId (1,2,3), (1,2,4) and (1,2,5) will all result in fresh invocations, although ids 1 and 2 are shared.

It's probably better to cache within `contentAPI.getTagInternalName`, which also has the benefit of hiding details of the cache implementation from consuming code.

Reply:
Done.